File: bindings/java/src/main/java/tech/turso/EncryptionOpts.java
package tech.turso;

import tech.turso.annotations.Nullable;

/**
 * Local encryption options for Turso database.
 *
 * <p>Encryption is an experimental feature. To use it, you must enable the "encryption"
 * experimental feature when opening the database.
 *
 * <p>Supported cipher algorithms:
 *
 * <ul>
 *   <li>aes128gcm (requires 16-byte/32-hex-char key)
 *   <li>aes256gcm (requires 32-byte/64-hex-char key)
 *   <li>aegis256 (requires 32-byte/64-hex-char key)
 *   <li>aegis256x2 (requires 32-byte/64-hex-char key)
 *   <li>aegis256x4 (requires 32-byte/64-hex-char key)
 *   <li>aegis128l (requires 16-byte/32-hex-char key)
 *   <li>aegis128x2 (requires 16-byte/32-hex-char key)
 *   <li>aegis128x4 (requires 16-byte/32-hex-char key)
 * </ul>
 *
 * <p>Example usage:
 *
 * <pre>{@code
 * EncryptionOpts encryptionOpts = new EncryptionOpts(
 *     "aegis256",
 *     "b1bbfda4f589dc9daaf004fe21111e00dc00c98237102f5c7002a5669fc76327"
 * );
 * }</pre>
 */
public final class EncryptionOpts {

  private final String cipher;
  private final String hexkey;

  /**
   * Creates a new encryption configuration.
   *
   * @param cipher The encryption cipher algorithm (e.g., "aegis256", "aes256gcm")
   * @param hexkey The encryption key as a hexadecimal string
   */
  public EncryptionOpts(String cipher, String hexkey) {
    if (cipher == null || cipher.isEmpty()) {
      throw new IllegalArgumentException("cipher must not be null or empty");
    }
    if (hexkey == null || hexkey.isEmpty()) {
      throw new IllegalArgumentException("hexkey must not be null or empty");
    }
    this.cipher = cipher;
    this.hexkey = hexkey;
  }

  /**
   * Gets the cipher algorithm.
   *
   * @return The cipher algorithm
   */
  public String getCipher() {
    return cipher;
  }

  /**
   * Gets the encryption key as a hexadecimal string.
   *
   * @return The hexadecimal encryption key
   */
  public String getHexkey() {
    return hexkey;
  }

  @Override
  public String toString() {
    return "EncryptionOpts{cipher='" + cipher + "', hexkey='***'}";
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    EncryptionOpts that = (EncryptionOpts) o;

    if (!cipher.equals(that.cipher)) return false;
    return hexkey.equals(that.hexkey);
  }

  @Override
  public int hashCode() {
    int result = cipher.hashCode();
    result = 31 * result + hexkey.hashCode();
    return result;
  }
}

File: bindings/dotnet/src/Turso/EncryptionOpts.cs
namespace Turso;

/// <summary>
/// Local encryption options for Turso database.
/// </summary>
/// <remarks>
/// Encryption is an experimental feature. To use it, you must enable the "encryption"
/// experimental feature when opening the database.
///
/// Supported cipher algorithms:
/// - aes128gcm (requires 16-byte/32-hex-char key)
/// - aes256gcm (requires 32-byte/64-hex-char key)
/// - aegis256 (requires 32-byte/64-hex-char key)
/// - aegis256x2 (requires 32-byte/64-hex-char key)
/// - aegis256x4 (requires 32-byte/64-hex-char key)
/// - aegis128l (requires 16-byte/32-hex-char key)
/// - aegis128x2 (requires 16-byte/32-hex-char key)
/// - aegis128x4 (requires 16-byte/32-hex-char key)
/// </remarks>
/// <example>
/// <code>
/// var encryptionOpts = new EncryptionOpts(
///     "aegis256",
///     "b1bbfda4f589dc9daaf004fe21111e00dc00c98237102f5c7002a5669fc76327"
/// );
/// </code>
/// </example>
public sealed class EncryptionOpts
{
    /// <summary>
    /// Gets the encryption cipher algorithm.
    /// </summary>
    public string Cipher { get; }

    /// <summary>
    /// Gets the encryption key as a hexadecimal string.
    /// </summary>
    public string Hexkey { get; }

    /// <summary>
    /// Creates a new encryption configuration.
    /// </summary>
    /// <param name="cipher">The encryption cipher algorithm (e.g., "aegis256", "aes256gcm")</param>
    /// <param name="hexkey">The encryption key as a hexadecimal string</param>
    /// <exception cref="ArgumentNullException">Thrown when cipher or hexkey is null</exception>
    /// <exception cref="ArgumentException">Thrown when cipher or hexkey is empty</exception>
    public EncryptionOpts(string cipher, string hexkey)
    {
        if (cipher == null)
            throw new ArgumentNullException(nameof(cipher));
        if (string.IsNullOrWhiteSpace(cipher))
            throw new ArgumentException("Cipher must not be empty", nameof(cipher));

        if (hexkey == null)
            throw new ArgumentNullException(nameof(hexkey));
        if (string.IsNullOrWhiteSpace(hexkey))
            throw new ArgumentException("Hexkey must not be empty", nameof(hexkey));

        Cipher = cipher;
        Hexkey = hexkey;
    }

    /// <summary>
    /// Returns a string representation of the encryption options with the key masked.
    /// </summary>
    public override string ToString()
    {
        return $"EncryptionOpts {{ Cipher = '{Cipher}', Hexkey = '***' }}";
    }

    public override bool Equals(object? obj)
    {
        if (obj is not EncryptionOpts other)
            return false;

        return Cipher == other.Cipher && Hexkey == other.Hexkey;
    }

    public override int GetHashCode()
    {
        return HashCode.Combine(Cipher, Hexkey);
    }
}
